datasets/za/pmg_legislators/crawler.py

```diff
@@ -0,0 +1,244 @@
+        if link.get("note") in ("website", "blog", "twitter", "facebook"):
+            person.add("website", url)
+
+    for contact_detail in person_data.pop("contact_details", []):
+        value = contact_detail.get("value")
+        if "email" == contact_detail.get("type"):
+            person.add("email", clean_emails(value))
+        if "address" == contact_detail.get("type"):
+            person.add("address", value)
+        if "postal_address" == contact_detail.get("type"):
+            person.add("address", value)
+
+    for membership in person_data.pop("memberships", []):
+        crawl_membership(context, person, membership, organizations, events)
+
+
+def crawl_membership(
+    context: Context,
+    entity: Entity,
+    membership: dict,
+    organizations: Dict[str, str],
+    events,
+) -> Optional[str]:
+    org_id = membership.get("organization_id")
+    org_name = organizations.get(org_id)
+    if not org_name:
+        # context.log.error(f"Organization with ID {org_id} not found.")
+        return None
+
+    role = membership.get("role")
+    if role is None:
+        # context.log.error("Role is missing in membership data.")
+        return None
+
+    # Creating the position property
+    position_property = post_summary(
+        org_name, role, [membership.get("start_date")], [membership.get("end_date")], []
+    )
+    entity.add("position", position_property)
+
+    # context.log.info(f"Processed membership for person ID: {person_id}")
+    position_label = f"{role.title()} of the {org_name}"
+
+    # Check if the role matches the positions of interest
+    if not any(re.findall(regex, position_label) for regex in POSITIONS_OF_INTEREST):
+        # context.log.info(f"Skipping role {role} not of interest.")
+        return None
+
+    # Initialize the topics list
+    topics = []
+
+    # Check if the role matches GOV_POSITIONS or LEG_POSITIONS
+    if any(re.findall(regex, position_label) for regex in GOV_POSITIONS):
+        topics.append("gov.national")
+
+    if any(re.findall(regex, position_label) for regex in LEG_POSITIONS):
+        topics.append("gov.legislative")
+
+    # If neither matches, we can skip this role
+    if not topics:
+        return None
+
+    position = h.make_position(
+        context,
+        position_label,
+        country="za",
+        topics=topics,
+    )
+
+    # Always PEP because filtered by known position label patterns
+    categorisation = categorise(context, position, is_pep=True)
+
+    period_id = membership.get("legislative_period_id")
+    period = events.get(period_id, {})
+    role = membership.pop("role", None)
+    role = role or period.get("name")
+
+    # If the role is not "member", we might want to log a warning.
+    # if role != "member":
+    #     context.log.warning("Unexpected role", role=role)
+
+    starts = [membership.get("start_date"), period.get("start_date")]
+    ends = [membership.get("end_date"), period.get("end_date")]
```

Comment:
we've got starts later than ends - here's an example for Naledi Pandor

```
                "name": [
                  "Member of the National Assembly"
                ],
                "country": [
                  "za"
                ]
              },
              "referents": [],
              "datasets": [
                "za_pmg_legislators"
              ],
              "first_seen": "2024-08-22T11:02:09",
              "last_seen": "2024-08-22T11:02:09",
              "last_change": "2024-08-22T11:02:09",
              "target": false
            }
          ],
          "startDate": [
            "2019-05-29"
          ],
          "endDate": [
            "2014-05-06"
          ],
          "status": [
            "ended"
          ],
          "holder": [
            "Q11043304"
          ]
```

Is that the data that's wrong, or an issue with combining dates from memberships and periods?

Are there many cases where membership isn't sufficient and we need to use period?

Does it help if we make it use either membership, or period, but not mix start/end from the two sources?

I'm a little nervous of using the period dates, to be honest, because we don't want to say someone was a member from january to december when they were actually a member from april to june, and it's just that the data couldn't tell us about their membership time.

What we do know is that they couldn't be a member later than the period end date.

We use the dates to determine whether they were a member within the PEP criteria (https://www.opensanctions.org/docs/pep/methodology/#types) so the most important thing is to include them if there's enough evidence to believe they were a member within the relevant period, and to to exclude them if there's evidence that they left the position long enough ago not to be considered a PEP.

We don't actually value a detailed chronology even though it looks cool to have it.

Reply:
this is actually about the dates we're passing to make_occupancy but applies here too.